core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/Version.java

```diff
@@ -0,0 +1,238 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2017 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.core.util;
+
+import javax.validation.constraints.NotNull;
+import java.util.Comparator;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.lang.Integer.parseInt;
+import static java.lang.String.format;
+import static java.util.regex.Pattern.compile;
+
+/**
+ * Version format:
+ * {MajorVersion}.{MinorVersion}.{BugFixVersion}.{HotFixVersion}-M{MilestoneVersion}-beta-{BetaVersion}-RC-SNAPSHOT.
+ * </p>
+ * Mandatory parts are only major, minor and bug fix versions.
+ *
+ * @see #toString()
+ * @see #VERSION
+ *
+ * @author Anatoliy Bazko
+ */
+public class Version implements Comparable<Version> {
+
+    private static final String MILESTONE_VERSION_SUFFIX = "-M";
+    private static final String BETA_VERSION_SUFFIX      = "-beta-";
+
+    private static final Pattern VERSION =
+            compile("^(?<major>0|[1-9]+[0-9]*)\\.(?<minor>0|[1-9]+[0-9]*)\\.(?<bugfix>0|[1-9]+[0-9]*)" +
+                    "([.](?<hotfix>0|[1-9]+[0-9]*))?" +
+                    "(" + MILESTONE_VERSION_SUFFIX + "(?<milestone>[1-9]+[0-9]*))?" +
+                    "(" + BETA_VERSION_SUFFIX + "(?<beta>[1-9]+[0-9]*))?" +
+                    "(?<rc>-RC)?" +
+                    "(?<snapshot>-SNAPSHOT)?$");
```

Comment:
I can imagine going even further and use something like next lines to eliminate usage of magic strings:
```
private static String MAJOR_GRP_NAME = "major";
...

private static String NON_ZERO_NUM_PART = "[1-9]+[0-9]*";
private static String ALLOW_ZERO_NUM_PART = "0|[1-9]+[0-9]*";

private static String MAJOR_GRP = "(?<major>" + ALLOW_ZERO_NUM_PART + ")";
private static String MINOR_GRP = "(?<minor>" + ALLOW_ZERO_NUM_PART + ")";
private static String BUGFIX_GRP = "(?<bugfix>" + ALLOW_ZERO_NUM_PART + ")";
private static String HOTFIX_GRP = "(\\.(?<hotfix>" + ALLOW_ZERO_NUM_PART + "))?";
private static String MILESTONE_GRP = "(" + MILESTONE_VERSION_SUFFIX + "(?<milestone>" + NON_ZERO_NUM_PART + "))?";
private static String BETA_GRP = "(" + BETA_VERSION_SUFFIX + "(?<beta>" + NON_ZERO_NUM_PART + "))?";
...
VERSION = Pattern.compile("^" + MAJOR_GRP + DOT + MINOR_GRP + DOT + BUGFIX_GRP + HOTFIX_GRP + MILESTONE_GRP + BETA_GRP + RC_GRP + SNAPSHOT_GRP);
```
And get groups by constants instead of magic strings too